### ISSUE TITLE
"grab-logs -s" leaks full WebDAV account informations

### ISSUE THREAD

**Issue by Alphakilo** [osmc/osmc#422]
According to your [contributing guidelines](https://osmc.tv/wiki/general/how-to-submit-a-useful-support-request/) the grab-logs script should not leak passwords:

> While generally the script tries to hide all passwords we know that certain people still are not comfortable with any information shared.

Welp. It doesn't hide the WebDAV account informations using `grab-logs -s` or `grab-logs -a`:
```
~# grab-logs -P -s
Logs created on: 2018-02-02 17:51:42 - (Uptime = 589.62)

SGkuGLGj  :  sources.xml


====================== sources.xml =================== SGkuGLGj
<sources>
    <programs>
        <default pathversion="1"></default>
    </programs>
    <video>
        <default pathversion="1"></default>
    </video>
    <music>
        <default pathversion="1"></default>
        <source>
            <name>Lokal</name>
            <path pathversion="1">/home/osmc/Music/</path>
            <allowsharing>true</allowsharing>
        </source>
        <source>
            <name>Nextcloud </name>
            <path pathversion="1">davs://$user:$password@$fqdn:443/$Path</path>
            <allowsharing>true</allowsharing>
        </source>
    </music>
    <pictures>
        <default pathversion="1"></default>
    </pictures>
    <files>
        <default pathversion="1"></default>
    </files>
</sources>

---------------------- sources.xml END --------------- SGkuGLGj
```
I discovered this during an -A run, thus leaking the Informations of my public avaible Server.
The Account is now disabled, passwords where changed.

Nevertheless this was not what I expected.

**Comment by samnazarko** [comment]
Thanks for the report. This is not intentional. The problem probably exists in Kodi upstream, and this would be the best place to report this issue so it can be fixed for all Kodi users. We will pick (and backport) any fixes for this.

**Comment by Alphakilo** [comment]
I'm sorry, but I have to object. `grablogs.py` does not exist in upstream.
The problem is not that the account informations are stored in cleartext (which is admittedly a problem in itself).

The quarrel I'm having is that `grablogs.py` picks up the `sources.xml` and not sanitizing the cleartext account information before running `dispatch_logs()` which breaks the premise given by the statement I cited above.
In fact, I haven't seen any sort of obfuscation going on in the script, let alone it "tries to hide all passwords"

I bet the same happens for (S)FTP sources as well.

**Comment by samnazarko** [comment]
Sorry -- I thought the issue was that Kodi had leaked these credentials in to Kodi.log. You should be using passwords.xml to define passwords; and they should not be in part of sources.xml. 

We will update the Wiki to make this clearer. We could introduce a regexy hack for trying to sanitise any logs that do contain credentials; but I'm not sure how effective this would be.

**Comment by Alphakilo** [comment]
> You should be using passwords.xml

Oooh okay. That's just the first time I heard about that. I've configured the sources via GUI.
Could you point me to some so documentation on the topic? kodi.wiki has nothing about it. And Google seems to only yield outdated examples.

From what I gathered it only covers SMB.

> We could introduce a regexy hack for trying to sanitise any logs that do contain credentials;

That'd be a great start, actually. I hacked together a RegEx over at regex101.com:
https://regex101.com/r/hJmqlf/4

The Python output looks something like this:
```
import re

regex = r"(?P<protocoll>(davs?|s?ftp|smb)(://))(?P<username>.+)(?P<cred_delim>(:))(?P<password>.+)(?P<delim>(@))"
```
I've fed it with some nasty strings, but they get encoded (URLEncode perhaps?) before hitting sources.xml. The regex101.com site I linked above includes some test cases.

I would provide a pull request, but I'm not proficient in Python and lack an OSMC dev environment.

**Comment by samnazarko** [comment]
Indeed, it may not cover webDAV. See https://forum.kodi.tv/showthread.php?tid=224876.

**Comment by samnazarko** [comment]
CC @KODeKarnage

**Comment by strugee** [comment]
I've submitted a draft patch to fix this issue upstream:  xbmc/xbmc#19217

**Comment by samnazarko** [comment]
If it’s backported to 19.5 then we will invoice it. We already have improvements to log sanitising which will come with Bullseye.

**Comment by strugee** [comment]
I'll ask upstream

**Comment by strugee** [comment]
https://github.com/xbmc/xbmc/pull/21475#event-8021508642 Matrix backport is merged
